Improve Printf implementation to use char arrays on GPU

aaronenyeshi opened the pull request:
Allocated a new buffer to store format strings used in printf calls.
Strings created on gpu will be copied into the string buffer on
host coherent memory, and will be printed after kernels. Users can
also write strings directly into arguments of printf call, as shown
in testcase.

TODO: We want to hide the printf implementation from user. Also we
want the printf buffer to be cyclical, so that we can reuse space
that is already printed. There is also a memory access fault that we
run into when the space used is close to buffer size, which needs
further investigation.

aaronenyeshi committed:
Improve Printf implementation to use char arrays on GPU

Allocated a new buffer to store format strings used in printf calls.
Strings created on gpu will be copied into the string buffer on
host coherent memory, and will be printed after kernels. Users can
also write strings directly into arguments of printf call, as shown
in testcase.
TODO: We want to hide the printf implementation from user. Also we
want the printf buffer to be cyclical, so that we can reuse space
that is already printed. There is also a memory access fault that we
run into when the space used is close to buffer size, which needs
further investigation.

scchan commented on include/hc_printf.hpp:
```diff
@@ -82,30 +93,69 @@ static inline void countArg(unsigned int& count, const T& t, const Rest&... rest
-  queue[offset].set(t);
+static inline PrintfError set_batch(PrintfPacket* queue, char* fbuffer, int offset, const T t) [[hc,cpu]] {
+  if (std::is_pointer<T>::value){
+    if (process_str_batch(queue, fbuffer, offset, (char*)t) == PRINTF_FORMAT_BUFFER_OVERFLOW)
```
This could be an issue here.  It assumes that any pointer passing to printf is a string, which is not the case since printf supports the %p format specifier.

scchan commented on include/hc_printf.hpp:
```diff
@@ -82,30 +93,69 @@ static inline void countArg(unsigned int& count, const T& t, const Rest&... rest
-  set_batch(queue, offset + 1, rest...);
+static inline PrintfError set_batch(PrintfPacket* queue, char* fbuffer, int offset, const T t, Rest... rest) [[hc,cpu]] {
+  if (std::is_pointer<T>::value){
+    if (process_str_batch(queue, fbuffer, offset, (char*)t) == PRINTF_FORMAT_BUFFER_OVERFLOW)
```
The check is too specific, it would be better to check if the return code != PRINTF_SUCCESS

scchan commented on include/hc_printf.hpp:
```diff
@@ -82,30 +93,69 @@ static inline void countArg(unsigned int& count, const T& t, const Rest&... rest
+static inline PrintfError set_batch(PrintfPacket* queue, char* fbuffer, int offset, const T t, Rest... rest) [[hc,cpu]] {
+  if (std::is_pointer<T>::value){
+    if (process_str_batch(queue, fbuffer, offset, (char*)t) == PRINTF_FORMAT_BUFFER_OVERFLOW)
+      return PRINTF_FORMAT_BUFFER_OVERFLOW;
```
This is too specific, it should just return the status from process_str_batch

scchan commented on include/hc_printf.hpp:
```diff
@@ -82,30 +93,69 @@ static inline void countArg(unsigned int& count, const T& t, const Rest&... rest
+    return PRINTF_FORMAT_BUFFER_OVERFLOW;
+  }
+  copy_n(&fbuffer[fb_offset], fstring, fstr_len);
+  fbuffer[fb_offset + fstr_len] = '\0';
```
It would be betterto do 
  copy_n(&fbuffer[fb_offset], fstring, **fstr_len + 1**);
then explicitly adding an extra '\0' at the end

aaronenyeshi committed:
Printf support for %p format specifier

We want to support %p format specifier for printf,
so we will check whether ptr is char* or const char*.
If users want to use %p for a char ptr, they will
need to first convert it into void*. Must add to doc
later.

aaronenyeshi commented:
@scchan I've added changes which address your comments, now we will check for char* and const char*. However for users who want to use %p specifier for their char arrays must first cast to void*.

aaronenyeshi committed:
Fix printf memory access fault

Seems that I used numelements * 8 as the size of
the buffer in the header but I didn't allocate the
same amount of space.

aaronenyeshi committed:
Hide printf format buffer inside printf header

Do not want the user to deal with the printf
format buffer, so we will maintain the pointer
inside the PrintfPacket buffer.
